WIP - Implement Get-Hash cmdlet

iSazonov opened the pull request:
It is first experimental implementation Get-Hash cmdlet for https://github.com/PowerShell/PowerShell-RFC/blob/master/2-Draft-Accepted/RFC0018-Get-StringHash.md

1. According to the RFC the type of Encoding parameter is `FileSystemCmdletProviderEncoding` but I had to use `EncodingConversion class` because the API contains useful methods and can be easily _enhanced_ in future (to support full set of codepages). Now multiple Utility cmdlets uses it. In contrast, a code using `FileSystemCmdletProviderEncoding` type is distributed in FileProvider and it's hard to reuse it although it is designated as the preferred (?) in https://github.com/PowerShell/PowerShell-RFC/blob/master/1-Draft/RFC0020-DefaultFileEncoding.md We also talked about that we should rationalize the use of encoding so `EncodingConversion class` looks like the best option.
2. Added `ValueFromPipelineByPropertyName` to `Path` parameter that is skipped in the RFC 0018.
3. ~~After the implementation we can really see that `Path` and `InputString` parameters have the same type and cannot be both `ValueFromPipeline`. Perhaps here we understand that it is better for UX to have both options and we should split on `Get-FileHash` and `Get-StringHash` still. Until concluding on this I haven't added pipeline tests.~~ We follow `Select-String` pattern as suggested by @rkeithhill.
4. ~~I wanted to implement `ValidateEncoding` attribute but was blocked so how to implement Tab Completion. I need help.~~ We should follow #4119.

iSazonov committed:
Implement Get-Hash cmdlet

rkeithhill commented:
RE #3, try changing the InputString type to `PSObject[]` and then convert each PSObject ToString before hashing.  That should get you around the pipeline binding issue.  This pattern in common for cmdlets that take both an InputObject and Path parameters like `Select-String`.

iSazonov commented:
@rkeithhill Thanks for workaround! The code looks not pretty. But I am ready to try this if community votes for this.

Can we introduce new `Path` type to resolve such problems? (This can be a lot of work)
```powershell
public class PSPath : PSObject
{
    string Path;
    public string ToString()
    {
        return Path;
    }
}
```

rkeithhill commented:
If you look at the SelectStringCommand impl - https://github.com/PowerShell/PowerShell/blob/02b5f357a20e6dee9f8e60e3adb9025be3c94490/src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs#L1009  it appears you use the `FilenfoToString` argument transform attribute on the `Path/LiteralPath` parameters.

iSazonov commented:
How do the attribute help to resolve conflict of parameter bindings?

rkeithhill commented:
Given that this attribute is private I've not had the chance to use it.  However, it makes `Select-String` work so I'm assuming here it should make `Get-Hash` work as well.  If you look at the parameter config for `Select-String` it is very similar to `Get-Hash`:
```
    Command: Microsoft.PowerShell.Utility/Select-String
    Set:    File *


Name                   Aliases      Position Mandatory Pipeline ByName Provider        Type
----                   -------      -------- --------- -------- ------ --------        ----
AllMatches             {A*}         Named    False     False    False  All             SwitchParameter
CaseSensitive          {Ca*}        Named    False     False    False  All             SwitchParameter
Context                {Co*}        Named    False     False    False  All             Int32[]
Encoding               {En*}        Named    False     False    False  All             String
Exclude                {Ex*}        Named    False     False    False  All             String[]
Include                {Inc*}       Named    False     False    False  All             String[]
List                   {Lis*}       Named    False     False    False  All             SwitchParameter
NotMatch               {N*}         Named    False     False    False  All             SwitchParameter
Path                   {}           1        True      False    True   All             String[]
Pattern                {Patt*}      0        True      False    False  All             String[]
Quiet                  {Q*}         Named    False     False    False  All             SwitchParameter
SimpleMatch            {S*}         Named    False     False    False  All             SwitchParameter


    Command: Microsoft.PowerShell.Utility/Select-String
    Set:    Object


Name                   Aliases      Position Mandatory Pipeline ByName Provider        Type
----                   -------      -------- --------- -------- ------ --------        ----
AllMatches             {A*}         Named    False     False    False  All             SwitchParameter
CaseSensitive          {Ca*}        Named    False     False    False  All             SwitchParameter
Context                {Co*}        Named    False     False    False  All             Int32[]
Encoding               {En*}        Named    False     False    False  All             String
Exclude                {Ex*}        Named    False     False    False  All             String[]
Include                {Inc*}       Named    False     False    False  All             String[]
InputObject            {Inp*}       Named    True      True     False  All             PSObject
List                   {Lis*}       Named    False     False    False  All             SwitchParameter
NotMatch               {N*}         Named    False     False    False  All             SwitchParameter
Pattern                {Patt*}      0        True      False    False  All             String[]
Quiet                  {Q*}         Named    False     False    False  All             SwitchParameter
SimpleMatch            {S*}         Named    False     False    False  All             SwitchParameter


    Command: Microsoft.PowerShell.Utility/Select-String
    Set:    LiteralFile


Name                   Aliases      Position Mandatory Pipeline ByName Provider        Type
----                   -------      -------- --------- -------- ------ --------        ----
AllMatches             {A*}         Named    False     False    False  All             SwitchParameter
CaseSensitive          {Ca*}        Named    False     False    False  All             SwitchParameter
Context                {Co*}        Named    False     False    False  All             Int32[]
Encoding               {En*}        Named    False     False    False  All             String
Exclude                {Ex*}        Named    False     False    False  All             String[]
Include                {Inc*}       Named    False     False    False  All             String[]
List                   {Lis*}       Named    False     False    False  All             SwitchParameter
LiteralPath            {PSPath, ... Named    True      False    True   All             String[]
NotMatch               {N*}         Named    False     False    False  All             SwitchParameter
Pattern                {Patt*}      0        True      False    False  All             String[]
Quiet                  {Q*}         Named    False     False    False  All             SwitchParameter
SimpleMatch            {S*}         Named    False     False    False  All             SwitchParameter
```
`Path` and `LiteralPath` are pipeline bound by property name and are of type string[].  `InputObject` is bound by value and is of type PSObject.  Somehow Select-String takes both string input to search through (bound to InputObject) and FileInfo objects from Get-ChildItem/Item.

daxian-dbw commented:
@TravisEz13 Could you please drive this PR?

iSazonov commented:
I'll continue on next week.

daxian-dbw commented:
@iSazonov No rush at all 😄 I was the assignee previously but I have too much on my plate right now and thus I reassigned Travis to this PR. That comment is just to let Travis know.

iSazonov commented:
I'm going to give my time and finish my work.

iSazonov committed:
Cleanup tests

iSazonov commented:
Ready for a code review.
I corrected the PR desciption and cleaned up test.
We have a conflict with #4119 for Encoding parameter.

iSazonov committed:
Add test for piping a path

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Hash.Tests.ps1:
```diff
@@ -0,0 +1,113 @@
+    }
+
+    Context "Paths tests" {
+        It "With '-Path': no file exist" {
```
Added the test .

iSazonov commented:
@TravisEz13 It is not test issue - it is temporary CI issue.

daxian-dbw commented:
@iSazonov himself is a maintainer now 😄

iSazonov commented:
I have no idea how to do it. Now I rename the last commit to restart CIs.

daxian-dbw commented:
@TravisEz13 Can we give Ilya permission to AppVeyor and Travis CI? I'm not sure if that requires one to be a member of PowerShell Org ...

lzybkr commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs:
```diff
@@ -78,6 +114,43 @@ protected override void BeginProcessing()
+                    {
+                        try
+                        {
+                            //if (str == null)
```
Delete commented out code, though if `str == null` is the only error condition, it's better to avoid exceptions and instead use the commented out code.

iSazonov replied:
Yes,  `str == null`  is the only error condition.
It seems `if` is lighter code then `try-catch`.

Fixed.

iSazonov committed:
Address code review

1. Replace GetFileHashCommand with GetHashCommand
2. Exclude an exception
3. Fix tests.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs:
```diff
@@ -8,10 +8,14 @@
+                                                                 StreamParameterSet
+                                                                 })]
+    [OutputType(typeof(StringHashInfo), ParameterSetName = new[] { StringHashParameterSet })]
     public class GetFileHashCommand : HashCmdletBase
```
Fixed.

lzybkr committed:
Tabs to spaces

lzybkr commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetHash.cs:
```diff
@@ -64,6 +68,38 @@ public String[] LiteralPath
+
+        /// <summary>
+        /// Encoding parameter
+        /// The Encoding of the 'InputString'
```
This comment is confusing because a `string` is always encoded in `Unicode`.
A better comment might be something like:

> the 'InputString' converted to the encoding if necessary

I do wonder if this parameter makes sense. In light of the approach in #4119 - the parameter becomes `FileEncoding`, which then becomes very confusing for this cmdlet - it applies to the string and not the file.

Given how uncommon it is to hash strings in other encodings, maybe it's fine to require the caller to convert to another encoding and this cmdlet also support hashing a byte array.

iSazonov replied:
I'll fix the comment.

About #4119. We have a conflict. We can have (1) split into two cmdlets - `Get-FileHash` and `Get-StringHash`, two parameters - `FileEncoding` and `StringEncoding`, or universal cmdlet `Get-Hash` with a special solution. If we want one cmdlet it would be good that it accept strings "as is".

lzybkr replied:
Maybe we should remove this parameter, but use UTF8 encoding.  Anyone wanting something else would write the string to a temp file and hash the file.

iSazonov replied:
Oh, the original idea was just to eliminate such workarounds.

iSazonov replied:
We could add an alias for  `FileEncoding` - `Encoding`.

iSazonov commented:
This is too old code. I will update it.
